SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,180 @@
+//===--- ExperimentalClosureSpecialization.swift ---------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===-----------------------------------------------------------------------===//
+
+/// AutoDiff Closure Specialization
+/// ----------------------
+/// This optimization performs closure specialization tailored for the patterns seen
+/// in Swift Autodiff. In principle, the optimization does the same thing as the existing
+/// closure specialization pass. However, it is tailored to the patterns of Swift Autodiff.
```

Comment:
Eventually we want this new pass replace the existing pass completely, right?

Reply:
Yes, that would be the goal. I was planning on creating this specific to Autodiff first and then extend it to general Swift as a fast follow-up. That is why I've created the `ClosureSpecializer` protocol below as well, which you've asked about below.

I _think_ (although I could be wrong) the callsite gathering logic for Autodiff is going to be different enough that we'd need to separate it out from the general Swift logic.